File: 백준/Gold/1300. K번째 수/K번째 수.py
"""
이분탐색

7번째 위치에 뭐가 있냐가 아니라 특정 수가 7번째 위치에 있냐로 찾기

mid보다 작은 수들의 개수와 작거나 같은 수들의 개수로 위치를 찾기
"""
import sys

input = lambda: sys.stdin.readline().strip()

N = int(input())
k = int(input())

answer = -1
s = 1
e = N * N
while True:
    mid = (s + e) // 2
    cnt_small_equal = 0
    cnt_small = 0
    for i in range(1, N+1):
        cnt_small_equal += min(N, mid // i)
        cnt_small += min(N, (mid-1) // i)
    if cnt_small < k <= cnt_small_equal:
        answer = mid
        break
    elif k <= cnt_small:
        e = mid - 1
    else:
        s = mid + 1

print(answer)

File: 백준/Gold/1300. K번째 수/K번째 수.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();
    private static int N, k;

    public static void main(String[] args) throws IOException {

        N  = Integer.parseInt(br.readLine());
        k = Integer.parseInt(br.readLine());

        long s = 1;
        long e = (long) N * N;

        long answer = 0;
        while (true) {
            long mid = (s + e) / 2;
            long cntSmallEqual = 0;
            long cntSmall = 0;
            for (int i = 1; i < N+1; i++) {
                cntSmallEqual += Math.min(N, mid / i);
                cntSmall += Math.min(N, (mid - 1) / i);
            }
            if (cntSmall < k && k <= cntSmallEqual) {
                answer = mid;
                break;
            } else if (k <= cntSmall) {
                e = mid - 1;
            } else {
                s = mid + 1;
            }
        }

        System.out.println(answer);

    } // end of main

} // end of class